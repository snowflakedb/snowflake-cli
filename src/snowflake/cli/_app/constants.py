# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Literal

PARAM_APPLICATION_NAME: Literal["snowcli"] = "snowcli"

# This is also defined on server side. Changing this parameter would require
# a change in https://github.com/snowflakedb/snowflake
INTERNAL_APPLICATION_NAME: Literal["SNOWFLAKE_CLI"] = "SNOWFLAKE_CLI"

# Authenticator types
AUTHENTICATOR_WORKLOAD_IDENTITY: Literal["WORKLOAD_IDENTITY"] = "WORKLOAD_IDENTITY"
AUTHENTICATOR_SNOWFLAKE_JWT: Literal["SNOWFLAKE_JWT"] = "SNOWFLAKE_JWT"
AUTHENTICATOR_USERNAME_PASSWORD_MFA: Literal[
    "username_password_mfa"
] = "username_password_mfa"
AUTHENTICATOR_OAUTH_AUTHORIZATION_CODE: Literal[
    "OAUTH_AUTHORIZATION_CODE"
] = "OAUTH_AUTHORIZATION_CODE"
