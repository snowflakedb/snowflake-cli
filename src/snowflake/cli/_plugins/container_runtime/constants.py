from snowflake.cli._plugins.container_runtime.utils import (
    ComputeResources,
    SnowflakeCloudType,
)

# SPCS specification constants
DEFAULT_CONTAINER_NAME = "main"
PAYLOAD_DIR_ENV_VAR = "MLRS_PAYLOAD_DIR"
RESULT_PATH_ENV_VAR = "MLRS_RESULT_PATH"
MEMORY_VOLUME_NAME = "dshm"
STAGE_VOLUME_NAME = "stage-volume"
STAGE_VOLUME_MOUNT_PATH = "/mnt/app"
USER_STAGE_VOLUME_NAME = "user-stage"
USER_STAGE_VOLUME_MOUNT_PATH = "/mnt/user-stage"

# Default container image information
DEFAULT_IMAGE_REPO = "/snowflake/images/snowflake_images"
DEFAULT_IMAGE_CPU = "st_plat/runtime/x86/runtime_image/snowbooks"
DEFAULT_IMAGE_GPU = "st_plat/runtime/x86/generic_gpu/runtime_image/snowbooks"
DEFAULT_IMAGE_TAG = "zzhu_vscode"
DEFAULT_ENTRYPOINT_PATH = "func.py"

# Percent of container memory to allocate for /dev/shm volume
MEMORY_VOLUME_SIZE = 0.3

# Ray port configuration
RAY_PORTS = {
    "HEAD_CLIENT_SERVER_PORT": "10001",
    "HEAD_GCS_PORT": "12001",
    "HEAD_DASHBOARD_GRPC_PORT": "12002",
    "HEAD_DASHBOARD_PORT": "12003",
    "OBJECT_MANAGER_PORT": "12011",
    "NODE_MANAGER_PORT": "12012",
    "RUNTIME_ENV_AGENT_PORT": "12013",
    "DASHBOARD_AGENT_GRPC_PORT": "12014",
    "DASHBOARD_AGENT_LISTEN_PORT": "12015",
    "MIN_WORKER_PORT": "12031",
    "MAX_WORKER_PORT": "13000",
}

# Node health check configuration
# TODO(SNOW-1937020): Revisit the health check configuration
ML_RUNTIME_HEALTH_CHECK_PORT = "5001"
ENABLE_HEALTH_CHECKS = "false"

# Job status polling constants
JOB_POLL_INITIAL_DELAY_SECONDS = 0.1
JOB_POLL_MAX_DELAY_SECONDS = 1

# Magic attributes
IS_MLJOB_REMOTE_ATTR = "_is_mljob_remote_callable"
RESULT_PATH_DEFAULT_VALUE = "mljob_result.pkl"

# Compute pool resource information
# TODO: Query Snowflake for resource information instead of relying on this hardcoded
#       table from https://docs.snowflake.com/en/sql-reference/sql/create-compute-pool
COMMON_INSTANCE_FAMILIES = {
    "CPU_X64_XS": ComputeResources(cpu=1, memory=6),
    "CPU_X64_S": ComputeResources(cpu=3, memory=13),
    "CPU_X64_M": ComputeResources(cpu=6, memory=28),
    "CPU_X64_L": ComputeResources(cpu=28, memory=116),
    "HIGHMEM_X64_S": ComputeResources(cpu=6, memory=58),
}
AWS_INSTANCE_FAMILIES = {
    "HIGHMEM_X64_M": ComputeResources(cpu=28, memory=240),
    "HIGHMEM_X64_L": ComputeResources(cpu=124, memory=984),
    "GPU_NV_S": ComputeResources(cpu=6, memory=27, gpu=1, gpu_type="A10G"),
    "GPU_NV_M": ComputeResources(cpu=44, memory=178, gpu=4, gpu_type="A10G"),
    "GPU_NV_L": ComputeResources(cpu=92, memory=1112, gpu=8, gpu_type="A100"),
}
AZURE_INSTANCE_FAMILIES = {
    "HIGHMEM_X64_M": ComputeResources(cpu=28, memory=244),
    "HIGHMEM_X64_L": ComputeResources(cpu=92, memory=654),
    "GPU_NV_XS": ComputeResources(cpu=3, memory=26, gpu=1, gpu_type="T4"),
    "GPU_NV_SM": ComputeResources(cpu=32, memory=424, gpu=1, gpu_type="A10"),
    "GPU_NV_2M": ComputeResources(cpu=68, memory=858, gpu=2, gpu_type="A10"),
    "GPU_NV_3M": ComputeResources(cpu=44, memory=424, gpu=2, gpu_type="A100"),
    "GPU_NV_SL": ComputeResources(cpu=92, memory=858, gpu=4, gpu_type="A100"),
}
CLOUD_INSTANCE_FAMILIES = {
    SnowflakeCloudType.AWS: AWS_INSTANCE_FAMILIES,
    SnowflakeCloudType.AZURE: AZURE_INSTANCE_FAMILIES,
}
