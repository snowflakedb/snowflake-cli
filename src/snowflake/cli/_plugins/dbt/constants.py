# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

RESULT_COLUMN_NAME = "SUCCESS"
OUTPUT_COLUMN_NAME = "STDOUT"

DBT_COMMANDS = [
    "build",
    "compile",
    "deps",
    "list",
    "parse",
    "run",
    "seed",
    "show",
    "snapshot",
    "test",
]

UNSUPPORTED_COMMANDS = [
    "clean",
    "clone",
    "debug",
    "docs",
    "init",
    "retry",
    "source",
]
