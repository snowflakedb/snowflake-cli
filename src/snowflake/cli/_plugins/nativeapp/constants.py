# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

SPECIAL_COMMENT_OLD = "GENERATED_BY_SNOWCLI"
SPECIAL_COMMENT = "GENERATED_BY_SNOWFLAKECLI"
ALLOWED_SPECIAL_COMMENTS = {SPECIAL_COMMENT, SPECIAL_COMMENT_OLD}
LOOSE_FILES_MAGIC_VERSION = "UNVERSIONED"

NAME_COL = "name"
COMMENT_COL = "comment"
OWNER_COL = "owner"
VERSION_COL = "version"
PATCH_COL = "patch"
CHANNEL_COL = "release_channel_name"
AUTHORIZE_TELEMETRY_COL = "authorize_telemetry_event_sharing"

INTERNAL_DISTRIBUTION = "internal"
EXTERNAL_DISTRIBUTION = "external"

DEFAULT_CHANNEL = "DEFAULT"
DEFAULT_DIRECTIVE = "DEFAULT"
MAX_VERSIONS_IN_RELEASE_CHANNEL = 2
