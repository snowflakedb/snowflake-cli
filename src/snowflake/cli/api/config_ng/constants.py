# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Constants for configuration system."""

from enum import Enum
from typing import Final, Literal


class ConfigSection(str, Enum):
    """Configuration section names."""

    CONNECTIONS = "connections"
    VARIABLES = "variables"
    CLI = "cli"
    CLI_LOGS = "cli.logs"
    CLI_FEATURES = "cli.features"

    def __str__(self) -> str:
        """Return the string value for backward compatibility."""
        return self.value


# Environment variable names
SNOWFLAKE_HOME_ENV: Final[str] = "SNOWFLAKE_HOME"

# Internal CLI parameters that should not be treated as connection parameters
INTERNAL_CLI_PARAMETERS: Final[set[str]] = {
    "enable_diag",
    "temporary_connection",
    "default_connection_name",
    "connection_name",
    "diag_log_path",
    "diag_allowlist_path",
    "mfa_passcode",
}

# Define Literal type for file source names
FileSourceName = Literal[
    "snowsql_config",
    "cli_config_toml",
    "connections_toml",
]

# Source names that represent file-based configuration sources
FILE_SOURCE_NAMES: Final[set[str]] = {
    "snowsql_config",
    "cli_config_toml",
    "connections_toml",
}
