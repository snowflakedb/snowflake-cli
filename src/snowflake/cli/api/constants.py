# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from pathlib import Path

TEMPLATES_PATH = Path(__file__).parent.parent / "templates"
DEPLOYMENT_STAGE = "deployments"
PYTHON_3_12 = (3, 12)


@dataclass(frozen=True)
class ObjectNames:
    cli_name: str
    sf_name: str
    sf_plural_name: str

    def __str__(self):
        return self.sf_name


class ObjectType(Enum):
    COMPUTE_POOL = ObjectNames("compute-pool", "compute pool", "compute pools")
    DATABASE = ObjectNames("database", "database", "databases")
    FUNCTION = ObjectNames("function", "function", "functions")
    INTEGRATION = ObjectNames("integration", "integration", "integrations")
    EXTERNAL_ACCESS_INTEGRATION = ObjectNames(
        "external-access-integration",
        "external access integration",
        "external access integrations",
    )
    # JOB = ObjectNames("job", "job", "jobs")
    NETWORK_RULE = ObjectNames("network-rule", "network rule", "network rules")
    NOTEBOOK = ObjectNames("notebook", "notebook", "notebooks")
    PROCEDURE = ObjectNames("procedure", "procedure", "procedures")
    ROLE = ObjectNames("role", "role", "roles")
    SCHEMA = ObjectNames("schema", "schema", "schemas")
    SERVICE = ObjectNames("service", "service", "services")
    SECRET = ObjectNames("secret", "secret", "secrets")
    STAGE = ObjectNames("stage", "stage", "stages")
    STREAM = ObjectNames("stream", "stream", "streams")
    STREAMLIT = ObjectNames("streamlit", "streamlit", "streamlits")
    TABLE = ObjectNames("table", "table", "tables")
    TASK = ObjectNames("task", "task", "tasks")
    USER = ObjectNames("user", "user", "users")
    WAREHOUSE = ObjectNames("warehouse", "warehouse", "warehouses")
    VIEW = ObjectNames("view", "view", "views")
    IMAGE_REPOSITORY = ObjectNames(
        "image-repository", "image repository", "image repositories"
    )
    GIT_REPOSITORY = ObjectNames("git-repository", "git repository", "git repositories")
    APPLICATION = ObjectNames("application", "application", "applications")
    APPLICATION_PACKAGE = ObjectNames(
        "application-package", "application package", "application packages"
    )

    def __str__(self):
        """This makes using this Enum easier in formatted string"""
        return self.value.cli_name


OBJECT_TO_NAMES = {o.value.cli_name: o.value for o in ObjectType}
UNSUPPORTED_OBJECTS = {
    ObjectType.APPLICATION.value.cli_name,
    ObjectType.APPLICATION_PACKAGE.value.cli_name,
}
SUPPORTED_OBJECTS = sorted(OBJECT_TO_NAMES.keys() - UNSUPPORTED_OBJECTS)

# Scope names here must replace spaces with '-'. For example 'compute pool' is 'compute-pool'.
VALID_SCOPES = ["database", "schema", "compute-pool"]

DEFAULT_SIZE_LIMIT_MB = 128

SF_REST_API_URL_PREFIX = "/api/v2"

PROJECT_TEMPLATE_VARIABLE_OPENING = "<%"
PROJECT_TEMPLATE_VARIABLE_CLOSING = "%>"

INIT_TEMPLATE_VARIABLE_OPENING = "<!"
INIT_TEMPLATE_VARIABLE_CLOSING = "!>"

SNOWPARK_SHARED_MIXIN = "snowpark_shared"

DEFAULT_ENV_FILE = "environment.yml"
DEFAULT_PAGES_DIR = "pages"
