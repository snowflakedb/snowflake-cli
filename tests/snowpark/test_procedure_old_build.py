# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from pathlib import Path
from textwrap import dedent
from unittest import mock
from unittest.mock import call

import pytest
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.errno import DOES_NOT_EXIST_OR_NOT_AUTHORIZED
from snowflake.cli.api.identifiers import FQN
from snowflake.connector import ProgrammingError

from tests_common import IS_WINDOWS

if IS_WINDOWS:
    pytest.skip("Requires further refactor to work on Windows", allow_module_level=True)


mock_session_has_warehouse = mock.patch(
    "snowflake.cli.api.sql_execution.SqlExecutionMixin.session_has_warehouse",
    lambda _: True,
)


@mock_session_has_warehouse
def test_deploy_function_no_procedure(runner, project_directory):
    with project_directory("empty_project"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
        )
    assert result.exit_code == 1
    assert (
        "No procedures or functions were specified in the project definition."
        in result.output
    )


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
):

    mock_om_describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures") as tmp:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ]
        )

    assert result.exit_code == 0, result.output
    mock_om_describe.return_value(
        [
            call(object_type=str(ObjectType.PROCEDURE), name="procedureName(string)"),
            call(object_type=str(ObjectType.PROCEDURE), name="test()"),
        ]
    )
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(tmp).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/ auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace procedure IDENTIFIER('MockDatabase.MockSchema.procedureName')(name string)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='hello'
            packages=()
            """
        ).strip(),
        dedent(
            """\
            create or replace procedure IDENTIFIER('MockDatabase.MockSchema.test')()
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='test'
            packages=()
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_with_external_access(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
):
    mock_om_describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    mock_om_show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]

    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedure_external_access") as project_dir:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ]
        )

    assert result.exit_code == 0, result.output
    mock_om_describe.assert_has_calls(
        [
            call(
                object_type=str(ObjectType.PROCEDURE),
                fqn=FQN.from_string("MockDatabase.MockSchema.procedureName(string)"),
            ),
        ]
    )
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace procedure IDENTIFIER('MockDatabase.MockSchema.procedureName')(name string)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.hello'
            packages=()
            external_access_integrations=(external_1, external_2)
            secrets=('cred'=cred_name, 'other'=other_name)
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_secrets_without_external_access(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    os_agnostic_snapshot,
):
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    mock_om_show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]

    with project_directory("snowpark_procedure_secrets_without_external_access"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 1, result.output
    assert result.output == os_agnostic_snapshot


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_fails_if_integration_does_not_exists(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    os_agnostic_snapshot,
):
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    mock_om_show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
    ]

    with project_directory("snowpark_procedure_external_access"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 1, result.output
    assert result.output == os_agnostic_snapshot


@mock.patch(
    "snowflake.cli._plugins.snowpark.commands._check_if_all_defined_integrations_exists"
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_fails_if_object_exists_and_no_replace(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    _,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
    os_agnostic_snapshot,
):
    mock_om_describe.return_value = mock_cursor(
        [
            ("packages", "[]"),
            ("handler", "hello"),
            ("returns", "string"),
        ],
        columns=["key", "value"],
    )
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy"])

    assert result.exit_code == 1
    assert result.output == os_agnostic_snapshot


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_replace_nothing_to_update(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
    caplog,
):
    mock_om_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "hello"),
                ("returns", "string"),
                ("imports", "dev_deployment/my_snowpark_project/app.py"),
            ],
            columns=["key", "value"],
        ),
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "test"),
                ("returns", "string"),
                ("imports", "dev_deployment/my_snowpark_project/app.py"),
                ("runtime_version", "3.10"),
            ],
            columns=["key", "value"],
        ),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {
            "object": "MockDatabase.MockSchema.test()",
            "status": "packages updated",
            "type": "procedure",
        },
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_replace_updates_single_object(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_om_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "hello"),
                ("returns", "string"),
                ("imports", "dev_deployment/my_snowpark_project/app.py"),
            ],
            columns=["key", "value"],
        ),
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "foo"),
                ("returns", "string"),
                ("imports", "dev_deployment/my_snowpark_project/app.zip"),
            ],
            columns=["key", "value"],
        ),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {
            "object": "MockDatabase.MockSchema.test()",
            "status": "definition updated",
            "type": "procedure",
        },
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_replace_creates_missing_object(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_om_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "hello"),
                ("returns", "string"),
                ("imports", "dev_deployment/my_snowpark_project/app.py"),
            ],
            columns=["key", "value"],
        ),
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {
            "object": "MockDatabase.MockSchema.test()",
            "status": "created",
            "type": "procedure",
        },
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_fully_qualified_name(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
):
    number_of_procedures_in_projects = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_procedures_in_projects
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedure_fully_qualified_name") as tmp_dir:
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="database error")


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_fully_qualified_name_duplicated_schema(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
):
    number_of_procedures_in_projects = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_procedures_in_projects
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedure_fully_qualified_name") as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path="snowpark.procedures.5.name",
            value="custom_schema.fqn_procedure_error",
        )
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="schema error")


@pytest.mark.parametrize(
    "parameter_type,default_value",
    [
        ("string", None),
        ("string", ""),
        ("int", None),
        ("variant", None),
        ("bool", None),
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_procedure_with_empty_default_value(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    parameter_type,
    default_value,
):
    mock_om_describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures") as project_dir:
        snowflake_yml = project_dir / "snowflake.yml"
        for param, value in [("type", parameter_type), ("default", default_value)]:
            alter_snowflake_yml(
                snowflake_yml,
                parameter_path=f"snowpark.procedures.0.signature.0.{param}",
                value=value,
            )
        result = runner.invoke(["snowpark", "deploy", "--format", "json"])

    default_value_json = default_value
    if default_value is None:
        default_value_json = "null"
    elif parameter_type == "string":
        default_value_json = f"'{default_value}'"

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": f"MockDatabase.MockSchema.procedureName(name {parameter_type} default {default_value_json})",
            "status": "created",
            "type": "procedure",
        },
        {
            "object": "MockDatabase.MockSchema.test()",
            "status": "created",
            "type": "procedure",
        },
    ]
