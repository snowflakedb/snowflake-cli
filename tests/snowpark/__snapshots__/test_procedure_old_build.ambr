# serializer version: 1
# name: test_deploy_procedure_fails_if_integration_does_not_exists
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Following external access integration does not exists in Snowflake:          |
  | external_2                                                                   |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_procedure_fails_if_object_exists_and_no_replace
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Following objects already exists. Consider using --replace.                  |
  | procedure: procedureName                                                     |
  | procedure: test                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_procedure_fully_qualified_name[database error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Database provided but name                                                   |
  | 'custom_database.custom_schema.fqn_procedure_error' is fully qualified name. |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_procedure_fully_qualified_name_duplicated_schema[schema error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Schema provided but name 'custom_schema.fqn_procedure_error' is fully        |
  | qualified name.                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_procedure_secrets_without_external_access
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | procedureName defined with secrets but without external integration.         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
